openpgp/packet/packet.go

```diff
@@ -525,6 +526,76 @@ func writeMPI(w io.Writer, bitLength uint16, mpiBytes []byte) (err error) {
 	return
 }
 
+func WritePaddedBigInt(w io.Writer, length int, X *big.Int) (n int, err error) {
+	bytes := X.Bytes()
+	n1, err := w.Write(make([]byte, length-len(bytes)))
+	if err != nil {
+		return n1, err
+	}
+	n2, err := w.Write(bytes)
+	if err != nil {
+		return n2, err
+	}
+	return (n1 + n2), err
+}
+
+// Minimum number of bytes to fit the curve coordinates. All
+// coordinates have to be 0-padded to this length.
+func mpiPointByteLength(curve elliptic.Curve) int {
+	return (curve.Params().P.BitLen() + 7) / 8
+}
+
+// Like elliptic.Unmarshal or ecdh.Unmarshal but you don't need to
+// know curve beforehand. It will read both uncompressed (0x4) and
+// compressed (0x40) points basing on header.
+func readPointMPI(r io.Reader) (X, Y *parsedMPI, err error) {
+	X = nil
+	Y = nil
+	var buf [2]byte
+	_, err = readFull(r, buf[0:2])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	mpiBitSize := int(buf[0])<<8 | int(buf[1])
+	mpiByteSize := (mpiBitSize + 7) / 8
+	r.Read(buf[:1])
+	header := buf[0]
+	if header == 0x4 {
+		coordLen := (mpiByteSize - 1) / 2
+
+		X = new(parsedMPI)
+		X.bytes = make([]byte, coordLen)
+		X.bitLength = uint16(coordLen * 8)
+		_, err = readFull(r, X.bytes)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		Y = new(parsedMPI)
+		Y.bytes = make([]byte, coordLen)
+		Y.bitLength = uint16(coordLen * 8)
+		_, err = readFull(r, Y.bytes)
+		if err != nil {
+			return nil, nil, err
+		}
```

Comment:
On last thought --- is it important to check that (X,Y) is on the curve here?  We do in the kbpgp implementation.